Mathlib/Data/Set/Finite/Basic.lean

```diff
@@ -693,6 +693,14 @@ theorem Finite.induction_on_subset {motive : ∀ s : Set α, s.Finite → Prop}
 @[deprecated (since := "2025-01-03")] alias Finite.induction_on' := Finite.induction_on_subset
 @[deprecated (since := "2025-01-03")] alias Finite.dinduction_on := Finite.induction_on
 
+theorem setOf_mem_list_eq_singleton_of_nodup {l : List α} (H : l.Nodup) {a : α} :
+    { x | x ∈ l } = {a} ↔ l = [a] := by
```

Comment:
This looks like it's provable without material on finiteness. You should also be able to generalise it to say l = replicate n a for some n > 0 without the nodup assumption

Reply:
```
theorem setOf_mem_list_eq_replicate {l : List α} {a : α} :
    { x | x ∈ l } = {a} ↔ ∃ n > 0, l = List.replicate n a := by
  simpa +contextual [Set.ext_iff, iff_iff_implies_and_implies, forall_and, List.eq_replicate_iff,
    List.length_pos_iff_exists_mem] using ⟨fun _ _ ↦ ⟨_, ‹_›⟩, fun x hx h ↦ h _ hx ▸ hx⟩

theorem setOf_mem_list_eq_singleton_of_nodup {l : List α} (H : l.Nodup) {a : α} :
    { x | x ∈ l } = {a} ↔ l = [a] := by
  constructor
  · rw [setOf_mem_list_eq_replicate]
    rintro ⟨n, hn, rfl⟩
    simp only [List.nodup_replicate] at H
    simp [show n = 1 by omega]
  · rintro rfl
    simp
```
here's what I mean - these things should be true just from list and set things, not finite set things